networkx/algorithms/tree/coding.py

```diff
@@ -0,0 +1,471 @@
+    -----
+    This function is *not* the inverse of :func:`to_nested_tuple`; the
+    only guarantee is that the rooted trees are isomorphic.
+
+    See also
+    --------
+    to_nested_tuple
+    from_prufer_sequence
+
+    Examples
+    --------
+    Sensible relabeling ensures that the nodes are labeled from the root
+    starting at 0::
+
+        >>> balanced = (((), ()), ((), ()))
+        >>> T = nx.from_nested_tuple(balanced, sensible_relabeling=True)
+        >>> list(T.edges())
+        [(0, 1), (0, 2), (1, 3), (1, 4), (2, 5), (2, 6)]
+
+    """
+
+    def _make_tree(sequence):
+        """Recursively creates a tree from the given sequence of nested
+        tuples.
+
+        This function employs the :func:`~networkx.tree.join` function
+        to recursively join subtrees into a larger tree.
+
+        """
+        # The empty sequence represents the empty tree, which is the
+        # (unique) graph with a single node. We mark the single node
+        # with an attribute that indicates that it is the root of the
+        # graph.
+        if len(sequence) == 0:
+            return nx.empty_graph(1)
+        # For a nonempty sequence, get the subtrees for each child
+        # sequence and join all the subtrees at their roots. After
+        # joining the subtrees, the root is node 0.
+        return nx.tree.join([(_make_tree(child), 0) for child in sequence])
+
+    # Make the tree and remove the `is_root` node attribute added by the
+    # helper function.
+    T = _make_tree(sequence)
+    if sensible_relabeling:
+        # Relabel the nodes according to their breadth-first search
+        # order, starting from the root node (that is, the node 0).
+        bfs_nodes = chain([0], (v for u, v in nx.bfs_edges(T, 0)))
+        labels = {v: i for i, v in enumerate(bfs_nodes)}
+        # We would like to use `copy=False`, but `relabel_nodes` doesn't
+        # allow a relabel mapping that can't be topologically sorted.
+        T = nx.relabel_nodes(T, labels)
+    return T
+
+
+def to_prufer_sequence_recursive(T):
+    """Returns the Prüfer sequence corresponding to the given tree.
+
+    ``T`` must be a tree, given as a NetworkX graph object with at least
+    two nodes.
+
+    """
+
+    def _sequence(T):
+        """Returns the Prüfer sequence corresponding to the given tree.
+
+        This is implemented as a separate function because the main
+        function neeeds to provide a copy of the tree here so that it
+        can be modified (by removing nodes) during the algorithm.
+
+        """
+        if len(T) == 2:
+            return []
+        try:
+            smallest = min(v for v in T if T.degree(v) == 1)
+        except ValueError:
+            raise NotATree('graph is not a tree because it has a cycle')
+        neighbor = arbitrary_element(T[smallest])
+        T.remove_node(smallest)
+        return [neighbor] + _sequence(T)
+
+    # TODO Is there a way of doing this without creating a copy of the graph?
```

Comment:
Yes, I thought of something like that when I first wrote this but decided to stick with the naive graph representation since it was simpler. Basically you really don't need the entire graph, you just need the parent of each node and the degree of each node, then update them manually as you pop nodes. It's probably a good idea to use that implementation: should be O(n) space instead of O(n + m), ignoring logarithmic factors.